tests/unit/test_shorthand.py

```diff
@@ -184,6 +204,38 @@ def test_parse(data, expected):
     actual = shorthand.ShorthandParser().parse(data)
     assert actual == expected
 
+class TestShorthandParserParamFile:
+    @patch('awscli.paramfile.compat_open')
+    @pytest.mark.parametrize(
+        'file_contents, data, expected',
+        (
+            ('file-contents123', 'Foo@=file://foo,Bar={Baz@=file://foo}', {'Foo': 'file-contents123', 'Bar': {'Baz': 'file-contents123'}}),
+            (b'file-contents123', 'Foo@=fileb://foo,Bar={Baz@=fileb://foo}', {'Foo': b'file-contents123', 'Bar': {'Baz': b'file-contents123'}}),
+            ('file-contents123', 'Bar@={Baz=file://foo}', {'Bar': {'Baz': 'file://foo'}}),
+            ('file-contents123', 'Foo@=foo,Bar={Baz@=foo}', {'Foo': 'foo', 'Bar': {'Baz': 'foo'}})
+        )
+    )
+    def test_paramfile(self, mock_compat_open, file_contents, data, expected):
+        mock_compat_open.return_value.__enter__.return_value.read.return_value = file_contents
+        result = shorthand.ShorthandParser().parse(data)
+        assert result == expected
+
+    @patch('awscli.paramfile.compat_open')
+    def test_paramfile_list(self, mock_compat_open):
+        f1_contents = 'file-contents123'
+        f2_contents = 'contents2'
+        mock_compat_open.return_value.__enter__.return_value.read.side_effect = [f1_contents, f2_contents]
```

Comment:
We should avoid changing the return type of `f.read()`, even if we're mocking it. Instead, since this test is just making sure paramfiles can be parsed in lists, we can just stick with 1 file used twice. And then we can also just add this to the parametrized test above:
```
('file-contents123', 'Foo@=[a, file://foo1, file://foo1]', {'Foo': ['a', 'file-contents123', 'file-contents123']})
```

Reply:
This code is a little different since its setting the value of `side_effect`, which let's you mock subsequent calls to the function via an iterable. So this array is actually patching the first read with element 0, and the second read with element 1.

I do feel there is value in having a test with 2 different files in a list. It verifies there is no bug of using the same file for every encounter of file:// and correctly resolves each element of the list independently.

Reply:
Ah I missed that. Okay that sounds good to me then